src/server/pkg/storage/fileset/fileset_test.go

```diff
@@ -1,274 +1,192 @@
+	"path"
+	"testing"
+
+	units "github.com/docker/go-units"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/require"
+	"github.com/pachyderm/pachyderm/src/server/pkg/dbutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/chunk"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/track"
+	"github.com/pachyderm/pachyderm/src/server/pkg/testutil"
+)
+
+const (
+	max         = 20 * units.MB
+	maxTags     = 10
+	testPath    = "test"
+	scratchPath = "scratch"
+)
+
+type testFile struct {
+	path  string
+	parts []*testPart
+}
+
+type testPart struct {
+	tag  string
+	data []byte
+}
+
+func appendFile(t *testing.T, w *Writer, path string, parts []*testPart) {
+	// Write content and tags.
+	err := w.Append(path, func(fw *FileWriter) error {
+		for _, part := range parts {
+			fw.Append(part.tag)
+			if _, err := fw.Write(part.data); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	require.NoError(t, err)
+}
+
+func overwritePart(t *testing.T, w *Writer, path string, tag string) {
+	panic("TODO")
+}
+
+func deletePart(t *testing.T, w *Writer, path string, tag string) {
+	err := w.Delete(path, tag)
+	require.NoError(t, err)
+}
+
+func writeFileSet(t *testing.T, s *Storage, fileSet string, files []*testFile) {
+	w := s.NewWriter(context.Background(), fileSet)
+	for _, file := range files {
+		appendFile(t, w, file.path, file.parts)
+	}
+	require.NoError(t, w.Close())
+}
+
+func checkFile(t *testing.T, f File, tf *testFile) {
+	r, w := io.Pipe()
+	eg := errgroup.Group{}
+	eg.Go(func() error {
+		return f.Content(w)
+	})
+	eg.Go(func() (retErr error) {
+		defer func() {
+			if retErr != nil {
+				r.CloseWithError(retErr)
+			} else {
+				r.Close()
+			}
+		}()
+		for _, part := range tf.parts {
+			actual := make([]byte, len(part.data))
+			_, err := io.ReadFull(r, actual)
+			if err != nil && err != io.EOF {
+				return err
+			}
+		}
+		return r.Close()
+	})
+	require.NoError(t, eg.Wait())
+}
+
+// newTestStorage creates a storage object with a test db and test tracker
+// both of those components are kept hidden, so this is only appropriate for testing this package.
+func newTestStorage(t *testing.T) *Storage {
+	db := dbutil.NewTestDB(t)
+	tr := track.NewTestTracker(t, db)
+	return NewTestStorage(t, db, tr)
+}
+
+func TestWriteThenRead(t *testing.T) {
+	ctx := testutil.NewTestContext(t)
+	storage := newTestStorage(t)
+	fileNames := index.Generate("abc")
+	files := []*testFile{}
+	for _, fileName := range fileNames {
+		var parts []*testPart
+		for _, tagInt := range rand.Perm(maxTags) {
+			tag := fmt.Sprintf("%08x", tagInt)
+			data := chunk.RandSeq(rand.Intn(max))
+			parts = append(parts, &testPart{
+				tag:  tag,
+				data: data,
+			})
+		}
+		files = append(files, &testFile{
+			path: "/" + fileName,
+		})
+	}
+
+	// Write out ten filesets where each subsequent fileset has the content of one random file changed.
```

Comment:
We should have a test that does what this comment describes. That is how this test worked previously. I am fine with leaving this test as a basic write / read check, then having a test called `TestWriteThenReadFuzz` which does what is described here. We also need a merge test.

Reply:
I'm going to hold of on the merge / compaction test until the new fileset API stuff is done.  There is actually a Merge method which takes a slice of fileset IDs and returns a single one.  So it will be pretty straightforward to write test vectors for that I think.